File: api/src/main/java/com/eventing/auth/AuthService.java
package com.eventing.auth;

import com.eventing.auth.dto.AuthResponse;
import com.eventing.auth.dto.LoginRequest;
import com.eventing.auth.dto.RegisterRequest;
import com.eventing.shared.config.AppConfig;
import com.eventing.shared.exception.ApiException;
import com.eventing.users.UserRepository;
import com.eventing.users.domain.User;
import io.quarkus.elytron.security.common.BcryptUtil;
import io.smallrye.jwt.build.Jwt;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import java.time.Duration;
import java.util.Set;

@ApplicationScoped
public class AuthService {

    @Inject
    UserRepository userRepository;

    @Inject
    AppConfig appConfig;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.email())) {
            throw ApiException.conflict("Email já cadastrado");
        }
        if (userRepository.existsByUsername(request.username())) {
            throw ApiException.conflict("Username já utilizado");
        }
        User user = new User();
        user.email = request.email();
        user.username = request.username();
        user.passwordHash = BcryptUtil.bcryptHash(request.password());
        user.role = "user";
        user.active = true;
        userRepository.persist(user);
        return buildResponse(user);
    }

    public AuthResponse login(LoginRequest request) {
        User user = userRepository.findByEmail(request.email())
                .filter(u -> u.active)
                .orElseThrow(ApiException::unauthorized);
        if (!BcryptUtil.matches(request.password(), user.passwordHash)) {
            throw ApiException.unauthorized();
        }
        return buildResponse(user);
    }

    private AuthResponse buildResponse(User user) {
        long lifespan = appConfig.getJwtLifespanSeconds();
        String token = Jwt.issuer(appConfig.getJwtIssuer())
                .subject(user.id.toString())
                .upn(user.email)
                .groups(Set.of(user.role))
                .expiresIn(Duration.ofSeconds(lifespan))
                .sign();
        return new AuthResponse(token, lifespan, user.id, user.username);
    }
}

File: app/src/services/authService.ts
import { type AuthResponse, type UserProfile } from '../types/api';
import { http, TOKEN_KEY } from './httpClient';

export async function register(
  email: string,
  username: string,
  password: string
): Promise<AuthResponse> {
  const auth = await http.post<AuthResponse>('/api/auth/register', {
    email,
    username,
    password,
  });
  localStorage.setItem(TOKEN_KEY, auth.token);
  return auth;
}

export async function login(email: string, password: string): Promise<AuthResponse> {
  const auth = await http.post<AuthResponse>('/api/auth/login', { email, password });
  localStorage.setItem(TOKEN_KEY, auth.token);
  return auth;
}

export async function me(): Promise<UserProfile> {
  return http.get<UserProfile>('/api/auth/me');
}

export function logout(): void {
  localStorage.removeItem(TOKEN_KEY);
}
